### DISCUSSION TITLE
Linter pass thorough Addons

### DISCUSSION THREAD

**Discussion by bgilmer77** [openhab/openhab-docs#2054] [Team Posts]
I have finished the bindings.  I am starting on the UI addons next.

**Comment by ThomDietrich** [comment]
Great work Brad!
Keep up the good work ;)

**Comment by Confectrician** [comment]
Wow. Big thanks. 👍

**Comment by bgilmer77** [comment]
UI addons complete (there was only one...).  Now starting on io addons.

**Comment by bgilmer77** [comment]
io addons done.  voice addons done.  linter pass through addons done?

@ThomDietrich what next?

**Comment by Confectrician** [comment]
all openhab addons or the openhab 2 addons?
openhab 1 should be checked also then.

**Comment by bgilmer77** [comment]
All OH2 I guess.  I will go find the OH1 addons next...

**Comment by Confectrician** [comment]
https://github.com/openhab/openhab-docs/projects/1

I have started a progress card for all locations and i will go through all repos and see what i can find.

**Comment by bgilmer77** [comment]
Very nice

**Comment by Confectrician** [comment]
Hey Brad, did you submit any additional pull requests already?
I want to go on with that porject test and start improving some linter readmes myself.

**Comment by Confectrician** [comment]
Hi again,

I have just pulled the latest oh2 addons master and my atom editor still showed some linter warnings, when i checked the first 2 bindings.
Did you verify your remark settings like mentioned by @ThomDietrich in this comment -> https://github.com/openhab/openhab2-addons/pull/2948#issuecomment-351656082 ?

![image](https://user-images.githubusercontent.com/7161177/34456919-e6aef1b8-eda2-11e7-9db7-55c6c819e47c.png)

**Comment by Confectrician** [comment]
And one reminder for me:

Code fences are done this way:
things -> java/perl (some syntaxes break the editor colors)
items/rules -> java
rules -> javascript (better sometimes)
sitemaps -> perl
everything else -> text
bash/shell -> shell

**Comment by bgilmer77** [comment]
Hi Jerome,
Yes, I have created some PRs on the OH1-addons side of things.  I am continuing to work through those next.

Okay on the linter rules.  Yes, this is exactly what I did, but perhaps I need to pull in the latest version.  I will do that before I work on the next set of PRs.

Okay on the code fences.  I was a little unclear before, but your note is super-clear.  Thank you.

**Comment by Confectrician** [comment]
Hi,

We have dicussed the code fences in some PR but i can't find it anymore without hours of searching.
So this rules are those that i reverse engineered now from the existing configuration docs.
Maybe they are not complete but at least we have a base to work with.

About openhab2-addons repo:
I have started a second run now and will take care of the left linter errors there.

**Comment by bgilmer77** [comment]
So I wonder about the rules...  Take a look at this PR - openhab/openhab1-addons/#5410.  In the Satel README.md, I followed your code fence rules above, but the highlighting in Atom for text that is outside of the fences is still affected.  Look toward the end of the file.  Is this how it should be?

**Comment by Confectrician** [comment]
I see no issue with the changed fences. just some sitemap demo that has not been improved.

I am more concerned about the lists.
The linter sometimes tells me to remove spaces and sometimes tells me to add 2 spaces and add a newline too.
I am waiting for an answer of @ThomDietrich in my corresponding oh2 addons PR for that.
This inconsequent behavior is confusing me a little bit.

**Comment by Confectrician** [comment]
OK i have found out something about the linting rules for lists:
http://www.cirosantilli.com/markdown-style-guide/#spaces-after-list-marker

The spaces and newlines are handled different, if the content fits in one line or not.

**Comment by bgilmer77** [comment]
Yes - I saw this in the style guide as well, and I believe the linter is working properly when it prompts you to change the spacing.  All of the files I have worked on I have followed what the linter says to do.  WDYT?

**Comment by Confectrician** [comment]
Yes. It seems to work correct.
I think we just have some disadvantages here compared to @ThomDietrich, who is dealing with this topic a bit longer than both of us.
So i would say, that will get easier for me/us to handle with some more experience. :)

**Comment by bgilmer77** [comment]
+1

**Comment by Confectrician** [comment]
Sorry for pushing here again and fpr poking @ThomDietrich again, but this time you have to explain me the behavior. 

I have no idea of whats going on here: (Have a look at the corresponding line numbers please)
![image](https://user-images.githubusercontent.com/7161177/34504587-b54daaae-f01f-11e7-9600-d95ebb04013d.png)

First lines are the following:

![image](https://user-images.githubusercontent.com/7161177/34504600-c901e862-f01f-11e7-99ce-be6924828611.png)

If i add a space, it changes the "add-warning" and also produces a "remove-warning".
I am changing spaces for 5min now and i am not able to remove the warnings.
Something must be wrong with the setting or the linter.

**Comment by Confectrician** [comment]
Seems that (remark-lint:list-item-bullet-indent) is causing trouble sometimes.

**Comment by bgilmer77** [comment]
What file are you working with?  I will take a look at it.

**Comment by Confectrician** [comment]
Development tutorial.
Contributing.md
  
I am currently fixing all linter errors in the dev tutorial.
Or better: I am trying to do so.

**Comment by ThomDietrich** [comment]
Hey guys!
Please forgive me for not being around as much. I didn't tell you ut after the release of openHAB 2.2 and openHABian 1.4 was through I concentrated on other more pressing matters in my professional life. I'll not be present as much in the next few months.

Regarding the linting rules: They are mostly following the style guide over at http://www.cirosantilli.com/markdown-style-guide
By that the recommendation for lists is a bit tricky and I know I've mentioned that before. I am still unsure if we should loosen the linter configuration a bit in that regard. In general I'd say that you don't have to satisfy every linter warning. And yes you could, I did not check the Contributing file but there probably is a solution. The important thing is to check for room for improvement. If a list is fine from your standpoint, just leave it be.

Don't hesitate to tune the linting rules!

One last thing: @Confectrician you wrote that `java` is best for codefences for rules. Java is fine in general but in some cases javascript is better. Sadly don't have an example to show right now.

**Comment by Confectrician** [comment]
Thanks for the clarificatoin @ThomDietrich and no need for excuses.
Voluntary work always has to stay aside, when professional life knocks at the door.
I would do it the same way, so that's totally ok.

Topic related:
My initial attempt was to get as less linter warnings as possible and remaining list warnings leave a huge amount warnings open.
But after your post i would change my way of dealing with this project.

I will try to improve the lists as they are and leave those warning open for now.
And i will investigate a linter rule change for the lists later then,  together with @bgilmer77.
It#s perfectly fine for me to get 80% progress now and work on the other 20% later or get rid of them with more flexible rules, as long as they are useful in the specific context.

One last thing:
Could you please have a quick look at https://github.com/openhab/openhab2-addons/pull/3055 and tell martinvw that the rules work correct, so that he knows that the PR can be merged, when i am ready.

**Comment by bgilmer77** [comment]
FWIW, I have been able to resolve almost all of the linter errors.  In the few cases where I could not, I have made a note in the PR, thinking we can go back and fix it later.

Also, I am getting ready to dive into the OH1.x binding addons next, but I have a little clean up to do on the 1.x actions I have already contributed.

**Comment by bgilmer77** [comment]
@Confectrician, the following fixes the linter errors associated with the lines you referenced...

```
- Your patch fixes spelling or grammar errors.
- Your patch is a single line change to documentation.
```
Basically change the "*" to a "-".
 
After that, the complaints about extra spaces required/not required go away.

**Comment by ThomDietrich** [comment]
Hey guys, while going through https://github.com/openhab/openhab2-addons/pull/3055 I found another linter exception:

https://github.com/openhab/openhab2-addons/pull/3055/files#r159464423

**Comment by Confectrician** [comment]
Hey Brad,

I have updated the project with some infos about the stuff i have already finished.
Maybe you could add a not about your progress too.

If you need more infos about how to edit the project, feel free to ask me.

BR
Jerome

**Comment by bgilmer77** [comment]
Hey Jerome,
Availability has been spotty since I have been having to focus on my paying job for a bit.  Yes - I will take a look at editing the project and updating it.  If I have questions I will ping you.

Brad

**Comment by bgilmer77** [comment]
Sigh... I finally got some time to do some testing, and found that I had not installed the linter correctly for the add-ons directory.  This is why you are seeing so many problems in my PRs.  Jeez - one more step up the learning curve.  So @Confectrician you are right to be going back through all of my submissions in the add-ons.

Jerome, since you are already working back through the OH2 addons, how about if I work my way back through the OH1 addons, starting with Actions.  I will then work through the OH1 Bindings.

I feel bad that I did not understand how the linter works, but that is the way it goes.

One question for you or for @ThomDietrich, now that I have installed the linter in the OH1 directory and have entered the linter installation stuff in `.gitignore`, the `.gitignore` file always shows up as modified when I get ready to do a commit, which means that I cannot use a -all flag.  I thought the way to get Git to ignore the `.gitignore` file would be to put an entry in the `gitignore` file which was `.gitignore` on its own line.  But that did not work.  How do you get git to ignore the `.gitignore` file?

**Comment by Confectrician** [comment]
Usually the .gitignore in .gitignore should work, but it doesn't work for me too at the moment.

**Comment by ThomDietrich** [comment]
Nooo 😅 don't do that guys :)
Ignoring an already committed file (.gitignore is already part of the repository) does not work. Entries in .gitignore are only honored for new files in the working tree.

Back to the issue: I suspect you want to ignore the package.json file during staging/commit? Simply said this is not possible. You'd either need to remove it or ignore it in gitignore. Both are not what we ultimately want, so the best option is really to manually ignore it. Yes that means not using `--all` - honestly I anyhow believe that this is a bad habit. It will eventually lead to files being committed you didn't want to commit. Try this instead:

```shell
git status
git add addons/*/*/README.md
git status # now use your brain :D
```

Btw great overall progress from both of you, thanks again!

**Comment by bgilmer77** [comment]
Yep - makes sense to me.  Since I am only working on one thing at a time, the -a flag makes a lot of sense.  If I were doing code development and touching a number of files while working on a project, then it would be different.

Hum - using brain is required.  I had no idea I would have to use my brain as part of this project ;-).
